design/history/exploration-reports/2020.10-data-model-numbers-and-codecs.md

```diff
@@ -0,0 +1,77 @@
+IPLD Data Model numbers and Ipld Codecs
+=======================================
+
+**Author**: Volker Mische ([@vmx])
+
+This is an exploration report about how to serialize [IPLD Data Model] numbers into bytes and back.
+
+
+Canonical representations
+-------------------------
+
+In IPLD it's desirable to have one single representation for every instance of the Data Model. This would lead to a bijective mapping between the Data Model and the encoded bytes. If you have a Data Model instance and you use a specific codec, it should always result in the same bytes, independent of the programming language used. That's important for content-addressing.
+
+
+### Conversion into the Data Model
+
+This looks like a hard problem, e.g. if you look at this IPFS issue with [DAG-CBOR] where there is a [different encoding between JavaScript and Go when numbers are integers]. Though in this issue, the problem is **not** the Data Model to Codec conversion. The problem is the **input data to Data Model conversion**. It took me a long time to figure that out.
+
+When I think of the IPLD Data Model I often think in terms of JSON and having numbers represented as text. I then think about how to represent those things as bytes. If you think about it that way, you have problems like: is `4251.00` and `4251` the same number? If yes, then it should be encoded the same way?
+
+From a Data Model perspective that's a non-issue. You have two kinds of numbers: [Integer] and [Float]. You can clearly distinguish between `Integer(4251)` and `Float(4251)`.
+
+
+### Serializing the Data Model
+
+
+#### Codec with multiple integer/float sizes
+
+Even within the Data Model there might be ambiguities, depending on Codec you use. If your Codec supports more than one integer type (which is often the case), you need to decide which one to choose. The obvious way is what CBOR in its [canonical representation] is doing:
+
+> Integers must be as small as possible.
+
+This means that a codec that supports 8-bit to 64-bit integers will encode `8472` always as `0x2118` and never as `0x00002118`.
+
+For floating point numbers (I always mean [binary IEEE-754 floating point numbers], others don't matter in modern day computing) the same issue arises. If your codec supports more than one floating point number type, e.g. 32-bit single precision and 64-bit double precision, a number like [`0.199951171875`] could be represented as `0x3e4cc000` (32-bit) or `0x3fc9980000000000` (64-bit). You can, just as with integers, require the smallest lossless representation of a float.
+
+Such a conversion to the smallest possible float makes sense for the sake of having a canonical representation, it doesn't make sense for space savings. Only about 0.00000002% of all the 64-bit floats can be represented as 32-bit floats. So it's not that likely that your data contains exactly those floats.
```

Comment:
Hah! that's a tiny percentage. I wonder if there's problems translating that into practice and whether "in practice, the types of floats generally represented by computer programs fall within .... certain bounds ...". Maybe in practice it's not as lop sided as this.

But this whole "smallest possible float" representation in CBOR sure is a major hassle and quite wasteful during encoding because you need to try encode+decode in 16, then 32 and then fall back to 64. With the most likely path, as you've suggested, being 64. Which is why I keep on coming back to this problem and wondering whether we're doing it wrong in JS or wrong in Go ... perhaps always-64 is the right choice?

Reply:
Using the "smallest possible float" makes sense for CBOR itself, but I don't think it makes sense for **DAG**-CBOR. I think we should switch to "always 64-bit". I think it would make encoders and decoders way simpler.